### PR TITLE
Refactor: arrange lints in `methods` module

### PR DESCRIPTION
**Pull request by TaKO8Ki**
This PR arranges methods lints so that they can be accessed more easily.
Basically, I refactored them following the instruction described in #6680.

changelog: Move lints in methods module into their own modules.

### CODE REVIEW

**Review by phansch — COMMENTED** [review]
I really like these refactors! Feel free to @ me again, when you're ready for a review =)

**Comment by TaKO8Ki** [comment]
@phansch 
Thank you!

**Comment by camsteffen** [comment]
You should probably split into a new PR every 10 or so lints for the reviewers' sanity 😄

**Comment by TaKO8Ki** [comment]
@camsteffen 
You’re right. I've already refactored most of lints in methods module. The others things to do is refactoring the rest of lints and moving some functions to `clippy_lints/src/utils/methods/`. I'm going to do them in a new PR!

**Comment by TaKO8Ki** [comment]
@phansch 
I'm ready for your review! ref: [comment above](https://github.com/rust-lang/rust-clippy/pull/6826#issuecomment-796473422)

**Comment by phansch** [comment]
@TaKO8Ki great! I will try to finish the review in the next 12 hours so that we don't run in too many conflicts with other PRs

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -4165,109 +2374,6 @@ fn is_bool(ty: &hir::Ty<'_>) -> bool {
...
-                let wildcards = format!("_{}", ", _".repeat(nb_wildcard - 1));
-                format!("{}<{}>", elements.join("::"), wildcards)
-            }
-        } else {
-            ty.to_string()
-        }
-    }
-}
-
 fn fn_header_equals(expected: hir::FnHeader, actual: hir::FnHeader) -> bool {
```
Comment (flip1995) [inline clippy_lints/src/methods/mod.rs:2377]: If those utils functions are used in more than one lint, please move it to a `methods/utils.rs` module. If it is only used in one lint, move it to the lint module. If it is only used in `mod.rs`, just leave it here.

Comment (flip1995) [reply clippy_lints/src/methods/mod.rs:2377]: A quick search would suggest, that this function should stay in `mod.rs`, but e.g. `derefs_to_slice` is used in multiple lint files and should therefore be moved to `methods/utils.rs`

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -1740,41 +1770,41 @@ impl<'tcx> LateLintPass<'tcx> for Methods {
...
+                                single_char_pattern::check(cx, expr, &args[pos]);
                             }
                         }
                     },
                     ty::Ref(..) if method_call.ident.name == sym::into_iter => {
-                        lint_into_iter(cx, expr, self_ty, *method_span);
+                        into_iter_on_ref::check(cx, expr, self_ty, *method_span);
                     },
                     _ => (),
                 }
```
Comment (Y-Nak) [inline clippy_lints/src/methods/mod.rs:1810]: These lines seem to do too many tasks.
I think checks that decide whether the lint is triggered should be moved to each module if duplications are not so much.

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (TaKO8Ki) [reply clippy_lints/src/methods/mod.rs:2377]: You're right. I think I should move some functions such as `derefs_to_slice` to `methods/utils.rs`, however this PR is too big. So, I'm going to fix this in a new PR.  ref [comment above](https://github.com/rust-lang/rust-clippy/pull/6826#issuecomment-796473422)

File: clippy_lints/src/methods/mod.rs (hunk 2 above)
Comment (TaKO8Ki) [reply clippy_lints/src/methods/mod.rs:1810]: Thank you for your suggestions! I think so too. However, this PR is too big and I have to commit many changes in order to move them to each module. So I'm going to do it in a new PR 👍 ref: https://github.com/rust-lang/rust-clippy/pull/6826#issuecomment-796473422

**Comment by TaKO8Ki** [comment]
**for all reviewers**

This PR is too big. So, if your suggestions are big, I'm going to accept them in a new PR. ref https://github.com/rust-lang/rust-clippy/pull/6826#issuecomment-796473422

**Comment by flip1995** [comment]
@TaKO8Ki sounds good to me! (And sorry for the merge conflict (-.-) )

**Comment by phansch** [comment]
I agree that we should keep further improvements out of this PR :+1:

**Comment by Y-Nak** [comment]
I agree that too.

**Comment by TaKO8Ki** [comment]
@flip1995 @phansch @Y-Nak 
I've just resolved conflicts!

**Comment by Y-Nak** [comment]
@TaKO8Ki Could you summarize the future works and reviews you received here in #6680 or a new issue?
It really helps us to track them easily.

**Comment by TaKO8Ki** [comment]
@Y-Nak 
I opened a tracking issue https://github.com/rust-lang/rust-clippy/issues/6886!

File: clippy_lints/src/methods/filter_map_flat_map.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+
+use super::FILTER_MAP;
+
+/// lint use of `filter_map().flat_map()` for `Iterators`
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'_>,
+    _filter_args: &'tcx [hir::Expr<'_>],
```
Comment (phansch) [inline clippy_lints/src/methods/filter_map_flat_map.rs:11]: We should review all new functions for unused arguments

File: clippy_lints/src/methods/iter_next_slice.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                    format!("{}.get({})", snippet_with_applicability(cx, caller_var.span, "..", &mut applicability), start_idx),
+                    applicability,
+                );
+            }
+        }
+    } else if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(caller_expr), sym::vec_type)
+        || matches!(
+            &cx.typeck_results().expr_ty(caller_expr).peel_refs().kind(),
+            ty::Array(_, _)
+        )
```
Comment (phansch) [inline clippy_lints/src/methods/iter_next_slice.rs:51]: This condition (not the body) could probably be extracted into its own function

File: clippy_lints/src/methods/iter_skip_next.rs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+
+use super::ITER_SKIP_NEXT;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &hir::Expr<'_>, skip_args: &[hir::Expr<'_>]) {
+    // lint if caller of skip is an Iterator
+    if match_trait_method(cx, expr, &paths::ITERATOR) {
+        if let [caller, n] = skip_args {
+            let hint = format!(".nth({})", snippet(cx, n.span, ".."));
```
Comment (phansch) [inline clippy_lints/src/methods/iter_skip_next.rs:12]: No need to have this as a separate variable

File: clippy_lints/src/methods/ok_expect.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+/// Given a `Result<T, E>` type, return its error type (`E`).
+fn get_error_type<'a>(cx: &LateContext<'_>, ty: Ty<'a>) -> Option<Ty<'a>> {
+    match ty.kind() {
+        ty::Adt(_, substs) if is_type_diagnostic_item(cx, ty, sym::result_type) => substs.types().nth(1),
+        _ => None,
+    }
+}
+
+/// This checks whether a given type is known to implement Debug.
+fn has_debug_impl<'tcx>(ty: Ty<'tcx>, cx: &LateContext<'tcx>) -> bool {
```
Comment (phansch) [inline clippy_lints/src/methods/ok_expect.rs:41]: Might be worth moving to utils

File: clippy_lints/src/methods/string_extend_chars.rs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+use rustc_span::symbol::sym;
+
+use super::STRING_EXTEND_CHARS;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &hir::Expr<'_>, args: &[hir::Expr<'_>]) {
+    let obj_ty = cx.typeck_results().expr_ty(&args[0]).peel_refs();
+    if is_type_diagnostic_item(cx, obj_ty, sym::string_type) {
```
Comment (phansch) [inline clippy_lints/src/methods/string_extend_chars.rs:12]: Returning early here or using `if_chain` would reduce some indentation in a lot of lints, like here.

**Review by phansch — APPROVED** [review]
Left some comments unrelated to this refactoring that can be fixed in later PRs
